In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import random
import sys
from itertools import chain

import numpy as np

from bounding_box_classifiers.obama_detector import train_helper, BasicConvNet

In [3]:
# data_dir = "data/bounding-box-obama-detector"
# data_dir = "data/obama-detector"
data_dir = "partitioned_data/held-out-author-obama"

static_params = {
    "seed": 1,
    "data_dir": data_dir,
    "use_cuda": True,
    "val_batch_size": 64,
    "epochs": 10,
    "log_interval": 10
}

variable_params = {
    "lr": [0.1, 0.01, 0.001, 0.001],
    "momentum": [0.5, 0.6, 0.7, 0.8, 0.9],
    "batch_size": [32, 64],
    "use_pretrained": [True, False],
    "feature_extract": [True]
}

In [ ]:
# Run random search
used_variable_params = set()
param_space_size = np.prod([len(p) for p in variable_params])
while len(used_variable_params) < param_space_size:
    curr_params = {k: random.choice(v) for k, v in variable_params.items()}
    curr_variable_params = tuple(chain.from_iterable(curr_params.items()))

    if curr_variable_params not in used_variable_params:
        print("Current variable parameters:", curr_variable_params)
        used_variable_params.add(curr_variable_params)
        curr_params.update(static_params)
        train_helper(**curr_params)

Current variable parameters: ('lr', 0.001, 'momentum', 0.8, 'batch_size', 64, 'use_pretrained', True, 'feature_extract', True)
Params to learn:
	 classifier.6.weight
	 classifier.6.bias
Train Epoch: 1 [0/11688 (0%)]	Loss: 0.732665
Train Epoch: 1 [640/11688 (5%)]	Loss: 0.762627
Train Epoch: 1 [1280/11688 (11%)]	Loss: 0.664646
Train Epoch: 1 [1920/11688 (16%)]	Loss: 0.636222
Train Epoch: 1 [2560/11688 (22%)]	Loss: 0.726927
Train Epoch: 1 [3200/11688 (27%)]	Loss: 0.641186
Train Epoch: 1 [3840/11688 (33%)]	Loss: 0.628081
Train Epoch: 1 [4480/11688 (38%)]	Loss: 0.645677
Train Epoch: 1 [5120/11688 (44%)]	Loss: 0.580163
Train Epoch: 1 [5760/11688 (49%)]	Loss: 0.717607
Train Epoch: 1 [6400/11688 (55%)]	Loss: 0.608062
Train Epoch: 1 [7040/11688 (60%)]	Loss: 0.679671
Train Epoch: 1 [7680/11688 (66%)]	Loss: 0.643773
Train Epoch: 1 [8320/11688 (71%)]	Loss: 0.669899
Train Epoch: 1 [8960/11688 (77%)]	Loss: 0.703831
Train Epoch: 1 [9600/11688 (82%)]	Loss: 0.696633
Train Epoch: 1 [10240/11688 (87%)]	L

Train Epoch: 9 [1920/11688 (16%)]	Loss: 0.603263
Train Epoch: 9 [2560/11688 (22%)]	Loss: 0.682621
Train Epoch: 9 [3200/11688 (27%)]	Loss: 0.582096
Train Epoch: 9 [3840/11688 (33%)]	Loss: 0.620722
Train Epoch: 9 [4480/11688 (38%)]	Loss: 0.675849
Train Epoch: 9 [5120/11688 (44%)]	Loss: 0.626464
Train Epoch: 9 [5760/11688 (49%)]	Loss: 0.578815
Train Epoch: 9 [6400/11688 (55%)]	Loss: 0.641438
Train Epoch: 9 [7040/11688 (60%)]	Loss: 0.764130
Train Epoch: 9 [7680/11688 (66%)]	Loss: 0.562004
Train Epoch: 9 [8320/11688 (71%)]	Loss: 0.555567
Train Epoch: 9 [8960/11688 (77%)]	Loss: 0.650984
Train Epoch: 9 [9600/11688 (82%)]	Loss: 0.581534
Train Epoch: 9 [10240/11688 (87%)]	Loss: 0.662203
Train Epoch: 9 [10880/11688 (93%)]	Loss: 0.522562
Train Epoch: 9 [11520/11688 (98%)]	Loss: 0.649402

Test set: Average loss: 0.8297, Accuracy: 589/1208 (49%)

Train Epoch: 10 [0/11688 (0%)]	Loss: 0.588331
Train Epoch: 10 [640/11688 (5%)]	Loss: 0.579984
Train Epoch: 10 [1280/11688 (11%)]	Loss: 0.591806
Train Epoc

Train Epoch: 7 [3200/11688 (27%)]	Loss: 0.814990
Train Epoch: 7 [3840/11688 (33%)]	Loss: 0.618571
Train Epoch: 7 [4480/11688 (38%)]	Loss: 0.736867
Train Epoch: 7 [5120/11688 (44%)]	Loss: 0.727935
Train Epoch: 7 [5760/11688 (49%)]	Loss: 0.754528
Train Epoch: 7 [6400/11688 (55%)]	Loss: 0.724779
Train Epoch: 7 [7040/11688 (60%)]	Loss: 0.666819
Train Epoch: 7 [7680/11688 (66%)]	Loss: 0.747419
Train Epoch: 7 [8320/11688 (71%)]	Loss: 0.733239
Train Epoch: 7 [8960/11688 (77%)]	Loss: 0.694287
Train Epoch: 7 [9600/11688 (82%)]	Loss: 0.701534
Train Epoch: 7 [10240/11688 (87%)]	Loss: 0.662282
Train Epoch: 7 [10880/11688 (93%)]	Loss: 0.730874
Train Epoch: 7 [11520/11688 (98%)]	Loss: 0.745724

Test set: Average loss: 0.8177, Accuracy: 601/1208 (50%)

Train Epoch: 8 [0/11688 (0%)]	Loss: 0.686833
Train Epoch: 8 [640/11688 (5%)]	Loss: 0.743604
Train Epoch: 8 [1280/11688 (11%)]	Loss: 0.758457
Train Epoch: 8 [1920/11688 (16%)]	Loss: 0.820033
Train Epoch: 8 [2560/11688 (22%)]	Loss: 0.694462
Train Epoch: 

Train Epoch: 5 [4480/11688 (38%)]	Loss: 0.683126
Train Epoch: 5 [5120/11688 (44%)]	Loss: 0.637745
Train Epoch: 5 [5760/11688 (49%)]	Loss: 0.611929
Train Epoch: 5 [6400/11688 (55%)]	Loss: 0.613342
Train Epoch: 5 [7040/11688 (60%)]	Loss: 0.646183
Train Epoch: 5 [7680/11688 (66%)]	Loss: 0.558990
Train Epoch: 5 [8320/11688 (71%)]	Loss: 0.522440
Train Epoch: 5 [8960/11688 (77%)]	Loss: 0.554695
Train Epoch: 5 [9600/11688 (82%)]	Loss: 0.613115
Train Epoch: 5 [10240/11688 (87%)]	Loss: 0.625073
Train Epoch: 5 [10880/11688 (93%)]	Loss: 0.707348
Train Epoch: 5 [11520/11688 (98%)]	Loss: 0.507291

Test set: Average loss: 0.8647, Accuracy: 600/1208 (50%)

Train Epoch: 6 [0/11688 (0%)]	Loss: 0.479861
Train Epoch: 6 [640/11688 (5%)]	Loss: 0.623230
Train Epoch: 6 [1280/11688 (11%)]	Loss: 0.645986
Train Epoch: 6 [1920/11688 (16%)]	Loss: 0.600517
Train Epoch: 6 [2560/11688 (22%)]	Loss: 0.706764
Train Epoch: 6 [3200/11688 (27%)]	Loss: 0.666043
Train Epoch: 6 [3840/11688 (33%)]	Loss: 0.657393
Train Epoch: 

Train Epoch: 2 [3520/11688 (30%)]	Loss: 28.101379
Train Epoch: 2 [3840/11688 (33%)]	Loss: 35.170197
Train Epoch: 2 [4160/11688 (36%)]	Loss: 31.050867
Train Epoch: 2 [4480/11688 (38%)]	Loss: 25.137188
Train Epoch: 2 [4800/11688 (41%)]	Loss: 34.951828
Train Epoch: 2 [5120/11688 (44%)]	Loss: 25.962360
Train Epoch: 2 [5440/11688 (46%)]	Loss: 21.661266
Train Epoch: 2 [5760/11688 (49%)]	Loss: 18.580339
Train Epoch: 2 [6080/11688 (52%)]	Loss: 17.401766
Train Epoch: 2 [6400/11688 (55%)]	Loss: 65.444336
Train Epoch: 2 [6720/11688 (57%)]	Loss: 22.138977
Train Epoch: 2 [7040/11688 (60%)]	Loss: 41.582588
Train Epoch: 2 [7360/11688 (63%)]	Loss: 43.318325
Train Epoch: 2 [7680/11688 (66%)]	Loss: 23.587004
Train Epoch: 2 [8000/11688 (68%)]	Loss: 33.492664
Train Epoch: 2 [8320/11688 (71%)]	Loss: 34.193253
Train Epoch: 2 [8640/11688 (74%)]	Loss: 42.282295
Train Epoch: 2 [8960/11688 (77%)]	Loss: 45.941383
Train Epoch: 2 [9280/11688 (79%)]	Loss: 33.565090
Train Epoch: 2 [9600/11688 (82%)]	Loss: 14.348059


Train Epoch: 6 [7360/11688 (63%)]	Loss: 29.676722
Train Epoch: 6 [7680/11688 (66%)]	Loss: 43.127609
Train Epoch: 6 [8000/11688 (68%)]	Loss: 43.687725
Train Epoch: 6 [8320/11688 (71%)]	Loss: 21.201185
Train Epoch: 6 [8640/11688 (74%)]	Loss: 19.028934
Train Epoch: 6 [8960/11688 (77%)]	Loss: 30.426050
Train Epoch: 6 [9280/11688 (79%)]	Loss: 13.756411
Train Epoch: 6 [9600/11688 (82%)]	Loss: 13.444500
Train Epoch: 6 [9920/11688 (85%)]	Loss: 45.323071
Train Epoch: 6 [10240/11688 (87%)]	Loss: 30.150652
Train Epoch: 6 [10560/11688 (90%)]	Loss: 15.340322
Train Epoch: 6 [10880/11688 (93%)]	Loss: 34.818565
Train Epoch: 6 [11200/11688 (96%)]	Loss: 28.505241
Train Epoch: 6 [11520/11688 (98%)]	Loss: 27.272429

Test set: Average loss: 23.8065, Accuracy: 603/1208 (50%)

Train Epoch: 7 [0/11688 (0%)]	Loss: 44.007198
Train Epoch: 7 [320/11688 (3%)]	Loss: 28.903721
Train Epoch: 7 [640/11688 (5%)]	Loss: 14.904368
Train Epoch: 7 [960/11688 (8%)]	Loss: 53.822994
Train Epoch: 7 [1280/11688 (11%)]	Loss: 95.80

Train Epoch: 10 [10880/11688 (93%)]	Loss: 24.169380
Train Epoch: 10 [11200/11688 (96%)]	Loss: 27.032333
Train Epoch: 10 [11520/11688 (98%)]	Loss: 44.454044

Test set: Average loss: 28.5581, Accuracy: 598/1208 (50%)

Current variable parameters: ('lr', 0.01, 'momentum', 0.5, 'batch_size', 64, 'use_pretrained', False, 'feature_extract', True)
Params to learn:
	 classifier.6.weight
	 classifier.6.bias
Train Epoch: 1 [0/11688 (0%)]	Loss: 0.759171
Train Epoch: 1 [640/11688 (5%)]	Loss: 0.920278
Train Epoch: 1 [1280/11688 (11%)]	Loss: 1.114845
Train Epoch: 1 [1920/11688 (16%)]	Loss: 1.539050
Train Epoch: 1 [2560/11688 (22%)]	Loss: 1.562010
Train Epoch: 1 [3200/11688 (27%)]	Loss: 1.879009
Train Epoch: 1 [3840/11688 (33%)]	Loss: 2.536064
Train Epoch: 1 [4480/11688 (38%)]	Loss: 0.871443
Train Epoch: 1 [5120/11688 (44%)]	Loss: 1.554739
Train Epoch: 1 [5760/11688 (49%)]	Loss: 1.495108
Train Epoch: 1 [6400/11688 (55%)]	Loss: 1.513451
Train Epoch: 1 [7040/11688 (60%)]	Loss: 1.646881
Train Epoch: 1 [


Test set: Average loss: 1.0552, Accuracy: 567/1208 (47%)

Train Epoch: 9 [0/11688 (0%)]	Loss: 1.563548
Train Epoch: 9 [640/11688 (5%)]	Loss: 1.768427
Train Epoch: 9 [1280/11688 (11%)]	Loss: 1.080617
Train Epoch: 9 [1920/11688 (16%)]	Loss: 1.853920
Train Epoch: 9 [2560/11688 (22%)]	Loss: 1.161129
Train Epoch: 9 [3200/11688 (27%)]	Loss: 1.656111
Train Epoch: 9 [3840/11688 (33%)]	Loss: 1.402049
Train Epoch: 9 [4480/11688 (38%)]	Loss: 1.025740
Train Epoch: 9 [5120/11688 (44%)]	Loss: 1.300987
Train Epoch: 9 [5760/11688 (49%)]	Loss: 1.655722
Train Epoch: 9 [6400/11688 (55%)]	Loss: 1.862823
Train Epoch: 9 [7040/11688 (60%)]	Loss: 2.275528
Train Epoch: 9 [7680/11688 (66%)]	Loss: 1.250341
Train Epoch: 9 [8320/11688 (71%)]	Loss: 1.311225
Train Epoch: 9 [8960/11688 (77%)]	Loss: 1.246810
Train Epoch: 9 [9600/11688 (82%)]	Loss: 0.957785
Train Epoch: 9 [10240/11688 (87%)]	Loss: 3.122931
Train Epoch: 9 [10880/11688 (93%)]	Loss: 1.422907
Train Epoch: 9 [11520/11688 (98%)]	Loss: 1.120456

Test set: Av

Train Epoch: 7 [640/11688 (5%)]	Loss: 0.617225
Train Epoch: 7 [1280/11688 (11%)]	Loss: 0.702526
Train Epoch: 7 [1920/11688 (16%)]	Loss: 0.734122
Train Epoch: 7 [2560/11688 (22%)]	Loss: 0.707829
Train Epoch: 7 [3200/11688 (27%)]	Loss: 0.783796
Train Epoch: 7 [3840/11688 (33%)]	Loss: 0.624301
Train Epoch: 7 [4480/11688 (38%)]	Loss: 0.744310
Train Epoch: 7 [5120/11688 (44%)]	Loss: 0.720350
Train Epoch: 7 [5760/11688 (49%)]	Loss: 0.734483
Train Epoch: 7 [6400/11688 (55%)]	Loss: 0.721269
Train Epoch: 7 [7040/11688 (60%)]	Loss: 0.672760
Train Epoch: 7 [7680/11688 (66%)]	Loss: 0.746219
Train Epoch: 7 [8320/11688 (71%)]	Loss: 0.713574
Train Epoch: 7 [8960/11688 (77%)]	Loss: 0.700751
Train Epoch: 7 [9600/11688 (82%)]	Loss: 0.720144
Train Epoch: 7 [10240/11688 (87%)]	Loss: 0.651490
Train Epoch: 7 [10880/11688 (93%)]	Loss: 0.709202
Train Epoch: 7 [11520/11688 (98%)]	Loss: 0.718963

Test set: Average loss: 0.7698, Accuracy: 584/1208 (48%)

Train Epoch: 8 [0/11688 (0%)]	Loss: 0.688110
Train Epoch: 

# Best results from Random Search
## Bounding Box Data
Basic CNN, both with 62% test accuracy:
```
('lr', 0.001, 'momentum', 0.9, 'batch_size', 32)  # epoch 2
('lr', 0.01, 'momentum', 0.6, 'batch_size', 32)  # epoch 5
```
With VGG:
```
('lr', 0.001, 'momentum', 0.7, 'batch_size', 32, 'use_pretrained', True, 'feature_extract', True)  # epoch 3
```


## Unfiltered Data
Both with 64% test accuracy:
```
('lr', 0.001, 'momentum', 0.8, 'batch_size', 32)  # epoch 10
('lr', 0.01, 'momentum', 0.7, 'batch_size', 64)  # epoch 7
```